src/pybind/mgr/cephadm/module.py

```diff
@@ -1682,24 +1659,38 @@ def add_host(self, spec):
             raise OrchestratorError('New host %s (%s) failed check: %s' % (
                 spec.hostname, spec.addr, err))
 
-        self.inventory[spec.hostname] = spec.to_json()
-        self._save_inventory()
+        self.inventory.add_host(spec)
         self.cache.prime_empty_host(spec.hostname)
         self.offline_hosts_remove(spec.hostname)
         self.event.set()  # refresh stray health check
         self.log.info('Added host %s' % spec.hostname)
         return "Added host '{}'".format(spec.hostname)
 
-    @async_completion
-    def remove_host(self, host):
-        # type: (str) -> str
-        """
-        Remove a host from orchestrator management.
+    def _verify_host_is_empty(self, host):
+        daemons = dict(self.cache.get_daemons_with_volatile_status())
+        daemons_on_host = daemons.get(host)
+        if daemons_on_host:
+            names = ', '.join(daemons_on_host.keys())
+            s = f'Daemons still on host `{host}`: {names}\n'
+            s += 'Please remove all daemons from this host, before removing the host.\n'
+            s += '(Add --force to ignore this message)'
```

Comment:
There would still be daemons running on the host that are reporting to the cluster. Forcefully removing the host takes away your ability to remove these daemons. Not sure if that's the right move.
A `--force` should also imply a de-registration of these daemons from the cluster.

Not sure _how_ this is done with the tools we have, but I wanted to raise awareness of this.

Reply:
yeah, my idea was to add a `ceph orch drain host0` that marks the host as _drain and then let serve() drain the host.